#914 Refactor: Luxon-Depenency, formatFromMillis()

CharalambosMakridakis opened the pull request:
Hello,

this PR is related to #914.

- Removed the luxon dependencies (types, luxon) and the related imports.
- Refactored the function formatFromMillis(), which used methods from luxon to a custom one without dependencies.
- Added tests for formatFromMillis()

Test passed:
![pass](https://github.com/cpvalente/ontime/assets/92108690/cd7fa2bd-fc21-4771-9862-cd126da96e53)

Edit:
I added the functionality test before changing the "luxon method" to be sure that my function works as intended.

cpvalente commented:
Thank you for this @CharalambosMakridakis 

Ontime does not really have concepts of date, all times are in milliseconds from midnight in UTC
From a quick glance, it was unclear to me if your code could cover the use cases, so I think the first approach is to approximate the tests to the application.

I have made a sandbox with some examples here, 
https://stackblitz.com/edit/stackblitz-starters-w82o8o?file=index.mjs

If we could use those test cases it would give us confidence that the change does not break expected behaviour to users

```ts
const time1 = 76211123; // 21:10:11.123
console.log('Testing 21:10:11.123');
console.log('=================================');
expectEqual(formatFromMillis(time1, 'S'), '123'); // milliseconds
expectEqual(formatFromMillis(time1, 's'), '11'); // seconds (no padding)
expectEqual(formatFromMillis(time1, 'ss'), '11'); // seconds (padding)
expectEqual(formatFromMillis(time1, 'm'), '10'); // minute (no padding)
expectEqual(formatFromMillis(time1, 'mm'), '10'); // minute (padding)
expectEqual(formatFromMillis(time1, 'h'), '9'); // hour - 12 (no padding)
expectEqual(formatFromMillis(time1, 'hh'), '09'); // hour - 12 (padding)
expectEqual(formatFromMillis(time1, 'H'), '21'); // hour - 24 (no padding)
expectEqual(formatFromMillis(time1, 'HH'), '21'); // hour - 24 (padding)
expectEqual(formatFromMillis(time1, 'HH:mm:ss'), '21:10:11');
expectEqual(formatFromMillis(time1, 'hh:mm a'), '09:10 PM');
```

```ts
const time2 = 29167345; // 08:06:07.345
expectEqual(formatFromMillis(time2, 'S'), '345'); // milliseconds
expectEqual(formatFromMillis(time2, 's'), '7'); // seconds (no padding)
expectEqual(formatFromMillis(time2, 'ss'), '07'); // seconds (padding)
expectEqual(formatFromMillis(time2, 'm'), '6'); // minute (no padding)
expectEqual(formatFromMillis(time2, 'mm'), '06'); // minute (padding)
expectEqual(formatFromMillis(time2, 'h'), '8'); // hour - 12 (no padding)
expectEqual(formatFromMillis(time2, 'hh'), '08'); // hour - 12 (padding)
expectEqual(formatFromMillis(time2, 'H'), '8'); // hour - 24 (no padding)
expectEqual(formatFromMillis(time2, 'HH'), '08'); // hour - 24 (padding)
expectEqual(formatFromMillis(time2, 'HH:mm:ss'), '08:06:07');
expectEqual(formatFromMillis(time2, 'hh:mm a'), '08:06 AM');
```

CharalambosMakridakis commented:
Hello @cpvalente,

i added the test cases and refactored the code according to your requirements mentioned in the previous comment.

Tests passed:
![pass](https://github.com/cpvalente/ontime/assets/92108690/4e720f0d-c7d2-4b10-9a34-877b578ca743)

cpvalente commented:
Nice @CharalambosMakridakis , really great work.

I believe you just need to rebuild the package-lock by running `pnpm install` and the tests should pass in pipeline

CharalambosMakridakis commented:
Done @cpvalente

cpvalente commented on packages/utils/src/date-utils/timeFormatting.ts:
```diff
@@ -71,11 +70,80 @@ export function removeSeconds(timer: string): string {
+  }
+
+  const date: Date = new Date(millis);
+  const options: Intl.DateTimeFormatOptions = {
+      year: 'numeric',
+      month: '2-digit',
+      day: '2-digit',
+      hour: '2-digit',
+      minute: '2-digit',
+      second: '2-digit',
+      hour12: false,
+      hourCycle: 'h23', // Ensures the hour is always two digits
+      timeZone: 'UTC'
+  };
+
+  const formattedDate: string = date.toLocaleString('en-US', options);
+  const milliseconds: string = date.getUTCMilliseconds().toString().padStart(3, '0');
+
+  // Extract date and time components.
+  const match = formattedDate.match(/(\d{2})\/(\d{2})\/(\d{4}), (\d{2}):(\d{2}):(\d{2})/);
+  if (!match) {
+      throw new Error("Date format mismatch or null result from regex match.");
+  }
+
+  const [_, month, day, year, hour24, minute, second] = match;
```
Nice!
This logic is well structured and pretty clever.

I wonder if the maintenance would be made easier by avoiding the regex and instead having explicit logic for the datasets. something like (untested)


```ts
  const date: Date = new Date(millis);
  const year = date.getUTCFullYear();
  const month = (date.getUTCMonth() + 1).toString().padStart(2, '0');
  const day = date.getUTCDate().toString().padStart(2, '0');
  const hour24 = date.getUTCHours().toString().padStart(2, '0');
  const minute = date.getUTCMinutes().toString().padStart(2, '0');
  const second = date.getUTCSeconds().toString().padStart(2, '0');
  const milliseconds = date.getUTCMilliseconds().toString().padStart(3, '0');

  const hour12 = (date.getUTCHours() % 12 || 12).toString();
```

We can probably remove `year` `month` and `day` since there is no expectation of us having to deal with it (app has no days)

CharalambosMakridakis replied:
I refactored the regex to your recommendation and removed `year` `month` and `day`

cpvalente reviewed (changes requested):
Thank you for your work @CharalambosMakridakis 

The code is well structured and I am impressed you could recreate the formatting in such a succinct manner

We should add a few more test cases to cover the second example I gave, this will expose that we forgot to account for padding in seconds and minutes

```ts
const time2 = 29167345; // 08:06:07.345
expectEqual(formatFromMillis(time2, 'S'), '345'); // milliseconds
expectEqual(formatFromMillis(time2, 's'), '7'); // seconds (no padding)
expectEqual(formatFromMillis(time2, 'ss'), '07'); // seconds (padding)
expectEqual(formatFromMillis(time2, 'm'), '6'); // minute (no padding)
expectEqual(formatFromMillis(time2, 'mm'), '06'); // minute (padding)
expectEqual(formatFromMillis(time2, 'h'), '8'); // hour - 12 (no padding)
expectEqual(formatFromMillis(time2, 'hh'), '08'); // hour - 12 (padding)
expectEqual(formatFromMillis(time2, 'H'), '8'); // hour - 24 (no padding)
expectEqual(formatFromMillis(time2, 'HH'), '08'); // hour - 24 (padding)
expectEqual(formatFromMillis(time2, 'HH:mm:ss'), '08:06:07');
expectEqual(formatFromMillis(time2, 'hh:mm a'), '08:06 AM');
```

I also wonder if we can simplify the logic to make it easier to maintain. 
Explicit calculations per dataset could help us in future to add and understand the implementation

Good work!

CharalambosMakridakis commented on packages/utils/src/date-utils/timeFormatting.ts:
```diff
@@ -71,11 +70,80 @@ export function removeSeconds(timer: string): string {
 export function formatFromMillis(millis: number, format: string): string {
-  return DateTime.fromMillis(millis).toUTC().toFormat(format);
+  if (millis < 0) {
+    throw new Error("Input `millis` can't be smaller than zero.");
```
I would say, that the best way forward is to return a falsy value (undefined), because wrapping every consumer in a try-catch would lead to a chore, even though it would be better for clean code.

CharalambosMakridakis commented on packages/utils/src/date-utils/timeFormatting.ts:
```diff
@@ -71,11 +70,80 @@ export function removeSeconds(timer: string): string {
+  const hour12 = ((parseInt(hour24) + 11) % 12 + 1);
+
+  const replacements: Record<string, string> = {
+      'yyyy': year,
+      'MM': month,
+      'dd': day,
+      'HH': hour24,
+      'mm': minute,
+      'ss': second,
+      'S': milliseconds,
+      'h': hour12.toString(), // Non-padded 12-hour format
+      'hh': hour12.toString().padStart(2, '0'), // Padded 12-hour format
+      'H': parseInt(hour24).toString(), // Non-padded 24-hour format
+      's': second,
+      'm': minute,
+      'a': parseInt(hour24) >= 12 ? 'PM' : 'AM'
```
I agree with you, I restructered it accordingly.

CharalambosMakridakis commented:
Thanks for pointing out!

Indeed the unpadded seconds and minutes weren't parsed correctly.
I added the other tests and fixed the problem.

CharalambosMakridakis commented:
Pls review again @cpvalente.

Thanks in advance.

cpvalente commented:
Thank you @CharalambosMakridakis , I just need to try and figure out what is going on with the automation.

CharalambosMakridakis commented:
Hello @cpvalente,
My pnpm version is:
![image](https://github.com/cpvalente/ontime/assets/92108690/37d6b652-b648-434b-b7b2-8363bf69635b)

Maybe too new? Should I downgrade to 8.x.x?

cpvalente commented:
> Maybe too new? Should I downgrade to 8.x.x?

yes please, we will eventually upgrade to pnpm 9, but we havent yet done so.

CharalambosMakridakis commented:
Hello @cpvalente,

could you please take a look?

Thank you.

cpvalente commented:
Thank you @CharalambosMakridakis , it seems that there is something strange about your lock file.
I am unsure how you ran the command, but the file has changed the version nearly every dependency in the project

If it is ok with you I will merge your code into an intermediate branch so that I can investigate.
I have already changed the target but you will need to resolve the conflicts, I would suggest just removing the commit for your pnpm-lock

CharalambosMakridakis commented:
Hi @cpvalente,

the dependency versions in the lockfile are indeed completely different from those in the main branch.
I replaced the lockfile with the one from the luxon branch, ran pnpm install and I think that should fix it now.

cpvalente reviewed (approved):
very good work, thank you! 🚀

cpvalente commented:
(@CharalambosMakridakis just need a small last push to solve the merge conflicts, sorry about this)

CharalambosMakridakis commented:
> (@CharalambosMakridakis just need a small last push to solve the merge conflicts, sorry about this)

Which merge conflicts? 

![image](https://github.com/cpvalente/ontime/assets/92108690/07b1b333-7129-45e4-ae2e-ba8eda7ebc12)

cpvalente commented:
> > (@CharalambosMakridakis just need a small last push to solve the merge conflicts, sorry about this)
> 
> 
> 
> Which merge conflicts? 
> 
> 
> 
> ![image](https://github.com/cpvalente/ontime/assets/92108690/07b1b333-7129-45e4-ae2e-ba8eda7ebc12)
> 
> 
> 
> ![image](https://github.com/cpvalente/ontime/assets/92108690/6e6a626a-dc41-45d3-95e5-042054220b6c)
> 
> 
> 
> 

😅...
